File: jambeth/jambeth-util/src/main/java/de/osthus/ambeth/proxy/AbstractSimpleInterceptor.java
package de.osthus.ambeth.proxy;

import java.lang.reflect.Method;

import net.sf.cglib.proxy.MethodInterceptor;
import net.sf.cglib.proxy.MethodProxy;
import de.osthus.ambeth.exception.RuntimeExceptionUtil;

public abstract class AbstractSimpleInterceptor implements MethodInterceptor
{
	public static final Method finalizeMethod;

	public static final Method equalsMethod;

	static
	{
		try
		{
			equalsMethod = Object.class.getDeclaredMethod("equals", Object.class);
			finalizeMethod = Object.class.getDeclaredMethod("finalize");
		}
		catch (Throwable e)
		{
			throw RuntimeExceptionUtil.mask(e);
		}
	}

	@Override
	public final Object intercept(Object obj, Method method, Object[] args, MethodProxy proxy) throws Throwable
	{
		if (finalizeMethod.equals(method))
		{
			// Do nothing. This is to prevent unnecessary exceptions in tomcat in REDEPLOY scenarios
			return null;
		}
		if (equalsMethod.equals(method) && args[0] == obj)
		{
			// Do nothing. This is to prevent unnecessary exceptions in tomcat in REDEPLOY scenarios
			return Boolean.TRUE;
		}
		try
		{
			return interceptIntern(obj, method, args, proxy);
		}
		catch (Throwable e)
		{
			throw RuntimeExceptionUtil.mask(e, method.getExceptionTypes());
		}
	}

	protected abstract Object interceptIntern(Object obj, Method method, Object[] args, MethodProxy proxy) throws Throwable;
}

File: ambeth/Ambeth.Util/ambeth/proxy/AbstractSimpleInterceptor.cs
using System;
using System.Reflection;
using De.Osthus.Ambeth.Annotation;
using De.Osthus.Ambeth.Threading;
using De.Osthus.Ambeth.Util;
#if !SILVERLIGHT
using Castle.DynamicProxy;
using System.Threading;
#else
using Castle.Core.Interceptor;
using Castle.DynamicProxy;
#endif

namespace De.Osthus.Ambeth.Proxy
{
    public abstract class AbstractSimpleInterceptor : IInterceptor
    {
        public static readonly MethodInfo equalsMethod = ReflectUtil.GetDeclaredMethod(false, typeof(Object), typeof(bool), "Equals", typeof(Object));

        public void Intercept(IInvocation invocation)
        {
            if (equalsMethod.Equals(invocation.Method) && Object.ReferenceEquals(invocation.Arguments[0], invocation.Proxy))
            {
                // Do nothing. This is to prevent unnecessary exceptions in tomcat in REDEPLOY scenarios
                invocation.ReturnValue = true;
                return;
            }
            InterceptIntern(invocation);
        }

        protected abstract void InterceptIntern(IInvocation invocation);
    }
}